**Coursebook: Data Wrangling and Group by Aggregation**
- Part 3 of Data Analytics Specialization
- Course Length: 12 hours
- Last Updated: April 2019
___

- Author: [Samuel Chan](https://github.com/onlyphantom)
- Developed by [Algoritma](https://algorit.ma)'s product division and instructors team

# Background

## Top-Down Approach 

The coursebook is part of the **Data Analytics Specialization** offered by [Algoritma](https://algorit.ma). It takes a more accessible approach compared to Algoritma's core educational products, by getting participants to overcome the "how" barrier first, rather than a detailed breakdown of the "why". 

This translates to an overall easier learning curve, one where the reader is prompted to write short snippets of code in frequent intervals, before being offered an explanation on the underlying theoretical frameworks. Instead of mastering the syntactic design of the Python programming language, then moving into data structures, and then the `pandas` library, and then the mathematical details in an imputation algorithm, and its code implementation; we would do the opposite: Implement the imputation, then a succinct explanation of why it works and applicational considerations (what to look out for, what are assumptions it made, when _not_ to use it etc).

## Learn-by-Building

This coursebook is intended for participants who have completed the preceding courses offered in the **Data Analytics Developer Specialization**. This is the third course, **Data Wrangling and Group By Aggregation**.

The coursebook focuses on:
- Stacking and Unstacking
- Working with MultiIndex DataFrames
- Reshaping your DataFrame with Melt
- Using Group By Effectively

At the end of this course is a Graded Asssignment section, where you are expected to apply all that you've learned on a new dataset, and attempt the given questions.

# Data Wrangling and Reshaping

In the previous two courses, we've got our hands on a few common techniques and learned how to explore data using `pandas` built-in methods. Specifically, we've  in the first and second part of this series how to use the following inspection, diagnostic and exploratory tools: 

**Data Inspection**
- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`
- Subsetting using `.loc`, `.iloc` and conditionals
---
**Diagnostic and Exploratory**
- Tables
- Cross-Tables and Aggregates
- Using `aggfunc` for aggregate functions
- Pivot Tables
- Working with DateTime
- Working with Categorical Data
- Duplicates and Missing Value Treatment

The first half of this course serves as an extension from the last. We'll pick up some new techniques to supplement our EDA toolset. Let us begin with reshaping techniques. 

In [1]:
from pandas_datareader import data
import pandas as pd

In [189]:
symbol = ['AAPL', 'FB', 'GOOGL']
source = 'yahoo'
start_date = '2018-01-01'
end_date = '2019-04-24'
stock = data.DataReader(symbol, source, start_date, end_date)
stock.head()

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   

Attributes                     Open                                Close  \
Symbols           GOOGL        AAPL          FB        GOOGL        AAPL   
Date                                                                       
2018-01-02  1053.020020  170.160004  177.679993  1053.020020  172.259995   
2018-01-03  1073.430054  172.529999  181.880005  1073.930054  172.229996   
2018-01-04  1094.260010  172.539993  184.899994  1097.089966  173.029999   
2018-01-05  1101.800049  173.440002  185.589996  1103.449951  175.000000   
2018-01-08  1110.000000  174.350006  187.199997  1111.000000  174.350006   

Attributes                               Volume                         \
Symbols             FB        GOOGL        AAPL          FB      GOOGL   
Date                                                                     
2018-01-02  181.419998  1073.209961  25555900.0  18151900.0  1588300.0   
2018-01-03  184.669998  1091.520020  29517900.0  16886600.0  1565900.0   
2018-01-04  184.330002  1095.760010  22434600.0  13880900.0  1302600.0   
2018-01-05  186.850006  1110.290039  23660000.0  13574500.0  1512500.0   
2018-01-08  188.279999  1114.209961  20567800.0  17994700.0  1232200.0   

Attributes   Adj Close                           
Symbols           AAPL          FB        GOOGL  
Date                                             
2018-01-02  168.987320  181.419998  1073.209961  
2018-01-03  168.957886  184.669998  1091.520020  
2018-01-04  169.742706  184.330002  1095.760010  
2018-01-05  171.675278  186.850006  1110.290039  
2018-01-08  171.037628  188.279999  1114.209961

If you do not have the `pandas_datareader` module installed, or if you're following along this coursebook without an active connection, you can instead load it from the serialized object I stored in your `data_cache` folder. 

Creating the DataFrame object by reading from `pickle`:
- `stock = pd.read_pickle('data_cache/stock')`

Serializing the DataFrame object to a byte stream using `pickle`:
- `stock.to_pickle('data_cache/stock')`

In [3]:
stock.to_pickle('data_cache/stock')
stock = pd.read_pickle('data_cache/stock')
stock.head()

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   

Attributes                     Open                                Close  \
Symbols           GOOGL        AAPL          FB        GOOGL        AAPL   
Date                                                                       
2018-01-02  1053.020020  170.160004  177.679993  1053.020020  172.259995   
2018-01-03  1073.430054  172.529999  181.880005  1073.930054  172.229996   
2018-01-04  1094.260010  172.539993  184.899994  1097.089966  173.029999   
2018-01-05  1101.800049  173.440002  185.589996  1103.449951  175.000000   
2018-01-08  1110.000000  174.350006  187.199997  1111.000000  174.350006   

Attributes                               Volume                         \
Symbols             FB        GOOGL        AAPL          FB      GOOGL   
Date                                                                     
2018-01-02  181.419998  1073.209961  25555900.0  18151900.0  1588300.0   
2018-01-03  184.669998  1091.520020  29517900.0  16886600.0  1565900.0   
2018-01-04  184.330002  1095.760010  22434600.0  13880900.0  1302600.0   
2018-01-05  186.850006  1110.290039  23660000.0  13574500.0  1512500.0   
2018-01-08  188.279999  1114.209961  20567800.0  17994700.0  1232200.0   

Attributes   Adj Close                           
Symbols           AAPL          FB        GOOGL  
Date                                             
2018-01-02  168.987320  181.419998  1073.209961  
2018-01-03  168.957886  184.669998  1091.520020  
2018-01-04  169.742706  184.330002  1095.760010  
2018-01-05  171.675278  186.850006  1110.290039  
2018-01-08  171.037628  188.279999  1114.209961

Create a DataFrame by subsetting only the `Close` columns. Name it `closingprice`. Then, use `.isna().sum()` to count the number of missing values in each of the columns present in `closingprice`.

If there are any missing values, use the `.fillna(method='ffill')` method to fill those missing values:

In [4]:
## Your code below


## -- Solution code

If you pay close attention to the index of `stock`, you may already realized by now that there are days where no records were present. 2018-01-01, 2018-01-06, and 2018-01-07 were absent from our DataFrame because they happen to fall on weekends.

While the trading hours of [different stock markets differ](https://www.maybank-ke.com.sg/markets/markets-listing/trading-hours/) (the NYSE for example open its market floor from 9.30am to 4pm five days a week), on weekends as well as federal holidays all stock exchanges are closed for business.

We can create (or recreate) the index by passing in our own values. In the following cell we created a date range and create the index using that new date range:

In [5]:
closingprice = stock['Close']
quarter1 = pd.date_range(start="2019-01-01", end="2019-03-31")
closingprice = closingprice.reindex(quarter1)
closingprice.head(8)

Symbols           AAPL          FB        GOOGL
2019-01-01         NaN         NaN          NaN
2019-01-02  157.919998  135.679993  1054.680054
2019-01-03  142.190002  131.740005  1025.469971
2019-01-04  148.259995  137.949997  1078.069946
2019-01-05         NaN         NaN          NaN
2019-01-06         NaN         NaN          NaN
2019-01-07  147.929993  138.050003  1075.920044
2019-01-08  150.750000  142.529999  1085.369995

Now use forward-fill to fill the `NA` values:

In [6]:
## Your code below


## -- Solution code

## `stack()` and `unstack()`

`stack()` stack the prescribed level(s) from columns to index and is particularly useful on DataFrames having a multi-level columns. It does so by "shifting" the columns to create new levels on its index. 

This is easier understood when we just see an example. Notice that `stock` has a 2-level column (Attributes and Symbols) and 1-level index (Date):

In [7]:
stock.head(8)

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   
2018-01-09  175.059998  188.800003  1118.439941  173.410004  187.100006   
2018-01-10  174.300003  187.889999  1112.780029  173.000000  185.630005   
2018-01-11  175.490005  188.399994  1114.849976  174.490005  187.380005   

Attributes                     Open                                Close  \
Symbols           GOOGL        AAPL          FB        GOOGL        AAPL   
Date                                                                       
2018-01-02  1053.020020  170.160004  177.679993  1053.020020  172.259995   
2018-01-03  1073.430054  172.529999  181.880005  1073.930054  172.229996   
2018-01-04  1094.260010  172.539993  184.899994  1097.089966  173.029999   
2018-01-05  1101.800049  173.440002  185.589996  1103.449951  175.000000   
2018-01-08  1110.000000  174.350006  187.199997  1111.000000  174.350006   
2018-01-09  1108.199951  174.550003  188.699997  1118.439941  174.330002   
2018-01-10  1103.979980  173.160004  186.940002  1107.000000  174.289993   
2018-01-11  1106.479980  174.589996  188.399994  1112.310059  175.279999   

Attributes                               Volume                         \
Symbols             FB        GOOGL        AAPL          FB      GOOGL   
Date                                                                     
2018-01-02  181.419998  1073.209961  25555900.0  18151900.0  1588300.0   
2018-01-03  184.669998  1091.520020  29517900.0  16886600.0  1565900.0   
2018-01-04  184.330002  1095.760010  22434600.0  13880900.0  1302600.0   
2018-01-05  186.850006  1110.290039  23660000.0  13574500.0  1512500.0   
2018-01-08  188.279999  1114.209961  20567800.0  17994700.0  1232200.0   
2018-01-09  187.869995  1112.790039  21584000.0  12393100.0  1340400.0   
2018-01-10  187.839996  1110.140015  23959900.0  10529900.0  1036700.0   
2018-01-11  187.770004  1112.050049  18667700.0   9588600.0  1121200.0   

Attributes   Adj Close                           
Symbols           AAPL          FB        GOOGL  
Date                                             
2018-01-02  168.987320  181.419998  1073.209961  
2018-01-03  168.957886  184.669998  1091.520020  
2018-01-04  169.742706  184.330002  1095.760010  
2018-01-05  171.675278  186.850006  1110.290039  
2018-01-08  171.037628  188.279999  1114.209961  
2018-01-09  171.018005  187.869995  1112.790039  
2018-01-10  170.978760  187.839996  1110.140015  
2018-01-11  171.949951  187.770004  1112.050049

When we stack the `stock` DataFrame, we shrink the number of levels on its column by one: `stock` now has 1-level column named `Attributes`: 

In [8]:
stock.stack().head(8)

Attributes                 High          Low         Open        Close  \
Date       Symbols                                                       
2018-01-02 AAPL      172.300003   169.259995   170.160004   172.259995   
           FB        181.580002   177.550003   177.679993   181.419998   
           GOOGL    1075.979980  1053.020020  1053.020020  1073.209961   
2018-01-03 AAPL      174.550003   171.960007   172.529999   172.229996   
           FB        184.779999   181.330002   181.880005   184.669998   
           GOOGL    1096.099976  1073.430054  1073.930054  1091.520020   
2018-01-04 AAPL      173.470001   172.080002   172.539993   173.029999   
           FB        186.210007   184.100006   184.899994   184.330002   

Attributes              Volume    Adj Close  
Date       Symbols                           
2018-01-02 AAPL     25555900.0   168.987320  
           FB       18151900.0   181.419998  
           GOOGL     1588300.0  1073.209961  
2018-01-03 AAPL     29517900.0   168.957886  
           FB       16886600.0   184.669998  
           GOOGL     1565900.0  1091.520020  
2018-01-04 AAPL     22434600.0   169.742706  
           FB       13880900.0   184.330002

`unstack()` does the opposite: it "shifts" the levels from index axis onto column axis. Try and create a stack DataFrame, and then apply `unstack` on the new DataFrame to see it return to the original shape:

In [9]:
## Your code below


## -- Solution code

### Knowledge Check

Which of the following statement is correct?

- [ ] `stack()` changes the DataFrame from wide to long
- [ ] `unstack()` changes the DataFrame from long to wide
- [ ] `unstack()` changes the DataFrame from wide to long

## Melt

Speaking of reshaping a DataFrame from wide format to long, another method that should be in your toolset is `melt()`. Consider the following DataFrame, which is created from `pandas` MultiIndex Slicers method, `.xs()`:

In [33]:
aapl = stock.xs('AAPL', level='Symbols', axis=1)
aapl.head()

Attributes        High         Low        Open       Close      Volume  \
Date                                                                     
2018-01-02  172.300003  169.259995  170.160004  172.259995  25555900.0   
2018-01-03  174.550003  171.960007  172.529999  172.229996  29517900.0   
2018-01-04  173.470001  172.080002  172.539993  173.029999  22434600.0   
2018-01-05  175.369995  173.050003  173.440002  175.000000  23660000.0   
2018-01-08  175.610001  173.929993  174.350006  174.350006  20567800.0   

Attributes   Adj Close  
Date                    
2018-01-02  168.987320  
2018-01-03  168.957886  
2018-01-04  169.742706  
2018-01-05  171.675278  
2018-01-08  171.037628

In [51]:
aapl.shape

(329, 6)

The DataFrame above is wide: it has 329 rows and 6 columns. The `melt()` function gathers all the columns into one and store the value corresponding to each column such that the resulting DataFrame has 329 * 6 = 1,974 rows, along with the identifier and values columns:

In [55]:
aapl_melted = aapl.melt()
aapl_melted.head()

Attributes       value
0       High  172.300003
1       High  174.550003
2       High  173.470001
3       High  175.369995
4       High  175.610001

In [53]:
aapl_melted.shape

(1974, 2)

We can optionally specify one or more columns to be identifier variables (`id_vars`), which treat all other columns as value variables (`value_vars`):

In [60]:
aapl.reset_index().melt(id_vars='Date')[325:333]

Date Attributes       value
325 2019-04-18       High  204.149994
326 2019-04-22       High  204.940002
327 2019-04-23       High  207.750000
328 2019-04-24       High  208.479996
329 2018-01-02        Low  169.259995
330 2018-01-03        Low  171.960007
331 2018-01-04        Low  172.080002
332 2018-01-05        Low  173.050003

#### Knowledge Check

Add a cell below, and fill the missing values in `aapl` using any method deemed appropriate:

In [75]:
march = pd.date_range(start="2018-03-01", end="2019-03-31")
aapl = stock.xs('AAPL', level='Symbols', axis=1)
aapl = aapl.reindex(march)
aapl.head(10)

Attributes        High         Low        Open       Close      Volume  \
2018-03-01  179.779999  172.660004  178.539993  175.000000  48802000.0   
2018-03-02  176.300003  172.449997  172.800003  176.210007  38454000.0   
2018-03-03         NaN         NaN         NaN         NaN         NaN   
2018-03-04         NaN         NaN         NaN         NaN         NaN   
2018-03-05  177.740005  174.520004  175.210007  176.820007  28401400.0   
2018-03-06  178.250000  176.130005  177.910004  176.669998  23788500.0   
2018-03-07  175.850006  174.270004  174.940002  175.029999  31703500.0   
2018-03-08  177.119995  175.070007  175.479996  176.940002  23774100.0   
2018-03-09  180.000000  177.389999  177.960007  179.979996  32185200.0   
2018-03-10         NaN         NaN         NaN         NaN         NaN   

Attributes   Adj Close  
2018-03-01  172.375214  
2018-03-02  173.567078  
2018-03-03         NaN  
2018-03-04         NaN  
2018-03-05  174.167923  
2018-03-06  174.020172  
2018-03-07  172.404770  
2018-03-08  174.286118  
2018-03-09  177.280518  
2018-03-10         NaN

In [73]:
## Your code below


## -- Solution code

# Group By

Reshaping data is an important component of any data wrangling toolkit as it allows the analyst to "massage" the data into the desired shape for further processing. 

Another equally important technique is the group by operation. Analysts having some experience with SQL or other data analysis toolsets (R's `tidyverse` for example) will find the group by operation a familiar strategy in many analysis-heavy workflow.

Consider the following DataFrame:

In [159]:
aapl = stock.xs('AAPL', level='Symbols', axis=1)
aapl = aapl.round(2)
aapl['Close_Diff'] = aapl['Close'].diff()
aapl['Weekday'] = aapl.index.weekday_name
aapl['Month'] = aapl.index.month_name()
aapl.tail()

Attributes    High     Low    Open   Close      Volume  Adj Close  Close_Diff  \
Date                                                                            
2019-04-17  203.38  198.61  199.54  203.13  28906800.0     203.13        3.88   
2019-04-18  204.15  202.52  203.12  203.86  24195800.0     203.86        0.73   
2019-04-22  204.94  202.34  202.83  204.53  19439500.0     204.53        0.67   
2019-04-23  207.75  203.90  204.43  207.48  23323000.0     207.48        2.95   
2019-04-24  208.48  207.05  207.36  207.16  17540600.0     207.16       -0.32   

Attributes    Weekday  Month  
Date                          
2019-04-17  Wednesday  April  
2019-04-18   Thursday  April  
2019-04-22     Monday  April  
2019-04-23    Tuesday  April  
2019-04-24  Wednesday  April

Pay special attention to how the `Close_Diff` column was created. It's the difference between the `Close` value of a stock price on a given day and the following day.

Supposed we want to compare the `Close_Diff` between each `Weekday`; On average, does Tuesday record a higher difference between the Close price of Apple stock compared to a Thursday?

In [162]:
aapl.groupby('Weekday').mean()

Attributes        High         Low        Open       Close        Volume  \
Weekday                                                                    
Friday      187.314848  183.676970  185.743636  185.425455  3.663593e+07   
Monday      187.599516  183.823065  185.762258  185.754516  3.276571e+07   
Thursday    187.768806  184.261940  185.871493  186.024179  3.297309e+07   
Tuesday     188.197910  184.630896  186.133284  186.539254  3.083892e+07   
Wednesday   188.078358  184.316418  186.125224  186.287612  3.231086e+07   

Attributes   Adj Close  Close_Diff  
Weekday                             
Friday      183.815758   -0.464848  
Monday      184.168871    0.000484  
Thursday    184.365373    0.084776  
Tuesday     184.882985    0.313485  
Wednesday   184.651343    0.584776

We can pass in multiple columns to the `.groupby()` call like so:

In [166]:
aapl.groupby(by=['Weekday', 'Month']).mean().head(15)

Attributes               High         Low        Open       Close  \
Weekday Month                                                       
Friday  April      180.185000  176.708333  179.333333  177.836667   
        August     216.312000  213.284000  214.188000  215.378000   
        December   165.062500  159.440000  164.212500  160.232500   
        February   169.937500  166.391250  168.485000  168.145000   
        January    165.321250  162.722500  163.827500  164.648750   
        July       191.972500  189.092500  190.817500  190.430000   
        June       189.152000  186.678000  188.320000  188.162000   
        March      180.962222  177.538889  179.157778  179.350000   
        May        187.942500  184.850000  185.790000  186.827500   
        November   194.312000  189.254000  192.166000  191.270000   
        October    223.185000  216.880000  220.585000  220.502500   
        September  224.852500  221.135000  223.292500  222.135000   
Monday  April      183.233333  180.016667  181.108889  182.016667   
        August     214.530000  211.552500  212.737500  212.835000   
        December   166.858000  162.068000  164.318000  164.586000   

Attributes               Volume   Adj Close  Close_Diff  
Weekday Month                                            
Friday  April      3.459397e+07  176.158333   -1.930000  
        August     3.106042e+07  213.570000    1.356000  
        December   5.525532e+07  159.547500   -4.430000  
        February   4.141038e+07  166.728750   -0.312500  
        January    3.417136e+07  162.655000    1.745000  
        July       1.867482e+07  188.297500   -0.200000  
        June       3.235132e+07  186.054000   -0.256000  
        March      3.399699e+07  178.185556    0.168889  
        May        2.954305e+07  184.560000    1.310000  
        November   4.515566e+07  190.310000   -4.420000  
        October    3.856388e+07  218.797500    0.937500  
        September  4.719880e+07  220.417500   -1.487500  
Monday  April      2.864951e+07  180.611111    1.060000  
        August     2.552682e+07  211.005000    0.520000  
        December   4.385782e+07  163.884000    0.684000

Before we going into some practice, let's look at some **group by** strategies when dealing with MultiIndex DataFrames. Consider the following DataFrame:

In [191]:
stock.stack().head(8)

Attributes                 High          Low         Open        Close  \
Date       Symbols                                                       
2018-01-02 AAPL      172.300003   169.259995   170.160004   172.259995   
           FB        181.580002   177.550003   177.679993   181.419998   
           GOOGL    1075.979980  1053.020020  1053.020020  1073.209961   
2018-01-03 AAPL      174.550003   171.960007   172.529999   172.229996   
           FB        184.779999   181.330002   181.880005   184.669998   
           GOOGL    1096.099976  1073.430054  1073.930054  1091.520020   
2018-01-04 AAPL      173.470001   172.080002   172.539993   173.029999   
           FB        186.210007   184.100006   184.899994   184.330002   

Attributes              Volume    Adj Close  
Date       Symbols                           
2018-01-02 AAPL     25555900.0   168.987320  
           FB       18151900.0   181.419998  
           GOOGL     1588300.0  1073.209961  
2018-01-03 AAPL     29517900.0   168.957886  
           FB       16886600.0   184.669998  
           GOOGL     1565900.0  1091.520020  
2018-01-04 AAPL     22434600.0   169.742706  
           FB       13880900.0   184.330002

On the Columns, there is 1 level. On the Index, there are 2 levels: `Date` and `Symbols` respectively. Supposed we want to groupby the second level of the axis (`Symbols`), we can pass in `axis=0` (referring to the row) and `level=1` (referring to the second level of index). 

The following groups it by the stock symbol and aggregate the values from the remaining levels using the `median`: 

In [194]:
stock.stack().groupby(level=1, axis=0).median()

Attributes         High          Low         Open        Close      Volume  \
Symbols                                                                      
AAPL         186.149994   182.559998   183.899994   184.160004  29855800.0   
FB           173.399994   168.160004   170.669998   171.160004  20750500.0   
GOOGL       1132.349976  1111.010010  1119.000000  1120.540039   1721100.0   

Attributes    Adj Close  
Symbols                  
AAPL         182.364288  
FB           171.160004  
GOOGL       1120.540039

# Learn-by-Building
## Graded Assignment 2
### Exploratory Data Analysis

1. Start from reading the data in using the following line of code:

```
pd.read_pickle('data_cache/stock')
```

2. Use `.diff()` to calculate the difference between a stock's adjusted closing price (`Adj Close`) and the following day. Recall that `.diff()` defaults to `periods=1`, so it calculates the difference between an element and the same element of the previous row.

3. Calculate the mean difference in `Adj Close` for each of `FB`, `GOOGL`, and `AAPL` stocks. Which have the highest average?

    - [ ] `AAPL`
    - [ ] `FB`
    - [ ] `GOOGL`


4. Start from reading the stock data by unpickling from `data_cache/stock`, then stack the DataFrame. From the stacked DataFrame, call `.groupby()` with the following parameters:
    - `axis=0` for the row
    - `level=1` for the innermost index (`Symbols` instead of `Date`)
    Chain a mean aggregation on the `groupby` operation. 

5. Using the output from (4), which has the lowest `Volume` on average? The volume is the number of shares that are transacted (buy/sell) during a given day.

    - [ ] `AAPL`
    - [ ] `FB`
    - [ ] `GOOGL`
    
6. Read `data_input/techcrunch.csv`, a dataset that stores fundraising rounds and amounts from startup companies of different categories around the US. Use `groupby` and perform a `median` aggregation. Which `category` raised the most amount in funding on average (use the `median`):

    - [ ] `mobile`
    - [ ] `cleantech`
    - [ ] `biotech`
    - [ ] `consulting`

7. Among all companies in San Francisco, which of the following are not among the top 5 most funded (highest `raisedAmt`) companies?
    - Hint: You may need to subset for companies where `city` is `San Francisco`, then group by `company` and `sum` on `raisedAmt` within each group. If it's helpful, you may consider chaining `sort_values()` and set `ascending=False` to get the DataFrame sorted by `raisedAmt` from largest to smallest
    
    - [ ] `OpenTable`
    - [ ] `Friendster`
    - [ ] `Facebook`
    - [ ] `Snapfish`

In [201]:
## Your code below


## -- Solution code